File: sql-like-dsl/haskell/orderDependent.hs
import Data.Function ((&))

data Query = Query
    { querySelect :: String
    , queryFrom :: String
    , queryWhere :: String
    , queryLimit :: Int
    }
    deriving (Show)

data Empty
data WithSelect
data WithFrom
data WithWhere

newtype PartialQuery a
    = PartialQuery Query
    deriving (Show)

newtype FullQuery
    = FullQuery Query
    deriving (Show)

emptyQuery :: PartialQuery Empty
emptyQuery =
    PartialQuery $
        Query
            { querySelect = ""
            , queryFrom = ""
            , queryWhere = ""
            , queryLimit = 0
            }

setSelect :: String -> PartialQuery Empty -> PartialQuery WithSelect
setSelect str (PartialQuery query) = PartialQuery $ query{querySelect = str}

setFrom :: String -> PartialQuery WithSelect -> PartialQuery WithFrom
setFrom from (PartialQuery query) = PartialQuery $ query{queryFrom = from}

setWhere :: String -> PartialQuery WithFrom -> PartialQuery WithWhere
setWhere str (PartialQuery query) =
    PartialQuery $
        query{queryWhere = str}

setLimit :: Int -> PartialQuery WithWhere -> FullQuery
setLimit limit (PartialQuery query) =
    FullQuery $
        query{queryLimit = limit}

myQuery :: FullQuery
myQuery =
    emptyQuery
        & setSelect "*"
        & setFrom "bobby_table"
        & setWhere "some_value >= 3"
        & setLimit 10

File: sql-like-dsl/ocaml/order_dependent.ml
type query = { select : string; from : string; where : string; limit : int }

(* Steps *)
type empty
type with_select
type with_from
type with_where

(* Outcome *)
type 'a partial_query = Partial of query
type full_query = Full of query

let empty_query : empty partial_query =
  Partial { select = ""; from = ""; where = ""; limit = 0 }

let set_select : string -> empty partial_query -> with_select partial_query =
 fun select (Partial query) -> Partial { query with select }

let set_from : string -> with_select partial_query -> with_from partial_query =
 fun from (Partial query) -> Partial { query with from }

let set_where : string -> with_from partial_query -> with_where partial_query =
 fun where (Partial query) -> Partial { query with where }

let set_limit : int -> with_where partial_query -> full_query =
 fun limit (Partial query) -> Full { query with limit }

let string_of_query (Full query) =
  let { select; from; where; limit } = query in
  Printf.sprintf "SELECT %s FROM %s WHERE %s LIMIT %d" select from where limit

[@@@ocamlformat "disable"]
let%test_unit _ =
let ( => ) = [%test_eq: Base.string] in
   let qry  : full_query =
    empty_query
    |> set_select "*"
    |> set_from "bobby_table"
    |> set_where "some_value >= 3"
    |> set_limit 10
   in
   string_of_query qry
     => "SELECT * FROM bobby_table WHERE some_value >= 3 LIMIT 10"
